F-1: Binary tree.

ValeryEMalakhov opened the pull request:
Implemented binary tree.
Developed and implemented a tree traversal algorithm.
Added tests.

ValeryEMalakhov committed:
Implemented binary tree.
The algorithm is developed.
Added tests.

ValeryEMalakhov committed:
Removed the temporary variable.
Removed the Option[].

ValeryEMalakhov committed:
Add empty and one-node tests.

ValeryEMalakhov committed:
Fix algorithm.

ValeryEMalakhov committed:
Changed while to tailrec.

btrofimov commented on src/main/scala/com/prb/f1/Node.scala:
```diff
@@ -0,0 +1,60 @@
+
+case class Node[T](
+    value: T,
+    var left: Option[Node[T]] = None,
```
Seems ou can use val here instead of var for left and right

ValeryEMalakhov replied:
Done. Tests are performed successfully.

ValeryEMalakhov committed:
Has added comments to methods.
Corrected the type of values in the foreach.
Changed the Node values to immutable.

ValeryEMalakhov committed:
Has added `specs2` tests.

ValeryEMalakhov committed:
Small fixes.

ValeryEMalakhov committed:
Redesigned the test using `Unit specification`.

ValeryEMalakhov committed:
Small fixes.

ValeryEMalakhov committed:
Node converter test added.

ValeryEMalakhov committed:
Deleted old jUnit tests.

ValeryEMalakhov committed:
Rewrite algorithm without `get`s.

ValeryEMalakhov committed:
Small fixes.

ValeryEMalakhov committed:
Rewrite with recTreeTraversal.

ValeryEMalakhov committed:
No `copy`, but still no `tailrec`.

ValeryEMalakhov committed:
Fix `tailRecTreeTraversal`.

ValeryEMalakhov committed:
Fix `tailRecTreeTraversal`.

ValeryEMalakhov committed:
Fix `tailRecTreeTraversal`.

ValeryEMalakhov committed:
Delete tree visualisation.
